snapcraft/internal/build_providers/_lxd/_lxd.py

```diff
@@ -0,0 +1,277 @@
+            "The command line interface, container names or lifecycle handling may "
+            "change in upcoming releases."
+        )
+        self._lxd_client = pylxd.Client()
+        self._container = None  # type: Optional[pylxd.models.container.Container]
+
+    def create(self) -> None:
+        """Create the LXD instance and setup the build environment."""
+        self.launch_instance()
+
+    def destroy(self) -> None:
+        """Destroy the instance, trying to stop it first."""
+        self._stop()
+
+    def mount_project(self) -> None:
+        if self._PROJECT_DEVICE_NAME in self._container.devices:
+            return
+
+        # Resolve the home directory
+        home_dir = (
+            self._run(command=["printenv", "HOME"], hide_output=True).decode().strip()
+        )
+
+        self._container.sync()
+
+        self._container.devices[self._PROJECT_DEVICE_NAME] = {
+            "type": "disk",
+            "source": self.project._project_dir,
+            "path": os.path.join(home_dir, "project"),
+        }
+
+        try:
+            self._container.save(wait=True)
+        except pylxd.exceptions.LXDAPIException as lxd_api_error:
+            raise errors.ProviderMountError(
+                provider_name=self._get_provider_name(), error_message=lxd_api_error
+            ) from lxd_api_error
+
+    def clean_project(self) -> bool:
+        was_cleaned = super().clean_project()
+        if was_cleaned:
+            if self._container is None:
+                self._container = self._lxd_client.containers.get(self.instance_name)
+            self._stop()
+            self._container.delete(wait=True)
+        return was_cleaned
+
+    def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
+        self._ensure_container_running()
+
+        # TODO: better handling of larger files.
+        try:
+            source_data = self._container.files.get(name)
+        except pylxd.exceptions.LXDAPIException as lxd_api_error:
+            raise errors.ProviderFileCopyError(
+                provider_name=self._get_provider_name(), error_message=lxd_api_error
+            )
+        else:
+            with open(destination, "wb") as destination_data:
+                destination_data.write(source_data)
+        if delete and self._container.files.delete_available:
+            self._container.files.delete(name)
+        if delete and not self._container.files.delete_available:
+            logger.warning("File deletion not supported by this LXD version.")
+
+    def shell(self) -> None:
+        self._run(command=["/bin/bash"])
+
+    def _ensure_container_running(self) -> None:
+        self._container.sync()
+        if self._container.status.lower() != "running":
+            raise errors.ProviderFileCopyError(
+                provider_name=self._get_provider_name(),
+                error_message="container is not running, it is current state is: {!r].".format(
```

Comment:
"it is current state is"?

What does this error mean, that the LXD instance died? What might cause this?

Reply:
good typo catch, could happen if someone stops the container or removes LXD mid run.